chain/network/src/config.rs

```diff
@@ -104,54 +107,75 @@ pub struct NetworkConfig {
 }
 
 impl NetworkConfig {
+    // Constructs and validates the config.
+    // TODO(gprusak): make the output immutable, to enforce the invariants
+    // checked during validation: either make it return an Arc, or add an Inner type,
+    // so that NetworkConfig dereferences to Inner.
```

Comment:
So you can't really make stuff immutable in Rust, without going against the grain of the language. If you have `x: &mut T`, you can always do `*x = T::new()`. 

Practically, I'd either: 

* would not worry about this at all
* make fields private & add a bunch of getters

Using deref to fake read-only fields would feel to me like misusing the language: I think it's more productive to focus on taking advantage of what can be experssed naturally in Rust, rather than trying to express every last bit of invariants no matter what. Basically, 80/20 correctness